defaulting-capi-clusters/README.md

```diff
@@ -0,0 +1,57 @@
+
+## Defaulting with webhooks
+
+Using mutating webhooks allows us to centralize the defaulting logic into webhooks.
+Simplified clients can then rely on defaulted responses from the Kubernetes API by utilizing the server side dry-run feature.
+
+The centralized defaulting logic additionally enables us to have no hardcoded values in any clients.
+
+### Simple defaulting flow
+
+We initially want to have a very simple defaulting flow which allows us to create Kubernetes clusters reliably.
+The cluster creation is focused on kubectl-gs without excluding any other tooling to be used.
+
+This is a high level overview of the cluster creation flow until the Custom Resources are applied:
+1. `User` triggers cluster templating with `kubectl-gs template cluster`
+2. `kubectl-gs` fetches upstream CAPI templates
+3. `kubectl-gs` **removes** values from CAPI templates which we want to default differently
+4. `kubectl-gs` returns minimal templates
+5. `User` either `kubectl apply`s minimal templates or runs `dry-run`
+6. `Admission Webhooks` default all missing values
+
+From this overview we can deduct a set requirements for our admission webhooks:
+- Webhooks must not have any side-effects (e.g. no additional objects are created or deleted)
+- Webhooks must be able to fully default a Custom Resource without context of other Custom Resources in the request (e.g. a `MachineDeployment` must be fully defaulted even if the `Cluster` CR is not known)
+- Webhooks must be able to determine which default values apply to the individual installation
+- `kubectl-gs` must be able to remove values from the CAPI templates efficiently
+
+### Source of truth for default values
+
+Defining the source of truth for default values will be hard in the future with different CAPI versions and other needs for configurability.
+We want to iterate faster at first and therefore a simple source of truth which might be inflexible is needed.
+Versioning of configuration as releases for cluster upgrades is out of scope for this initial iteration.
+
+Webhooks can utilize the already existing configuration options through `config-controller` to have access to global and installation specific defaults.
+It will be necessary to package webhooks (such as `Kyverno policies`) as apps which have their configuration managed through the [config](https://github.com/giantswarm/config) repository.
+
+The workflow for a developer to add a new default configuration is then the following:
+1. `Developer` adds logic to new defaulting (e.g. a `kyverno policy`).
+2. `Developer` adds templating to the app (e.g. [kyverno policies for aws](https://github.com/giantswarm/kyverno-policies/blob/main/helm/policies-aws/values.yaml) with local defaults in `values.yaml`.
+3. `Developer` adds global and installation specific defaults through the [config](https://github.com/giantswarm/config) repository.
```

Comment:
When should someone store the value in `kyverno-policies` as opposed to `config`?

Reply:
Maybe weird phrasing: You need to make something configurable in the `values.yaml` to be able to overwrite it through the `config` repository.

Reply:
Gotcha -- I think the confusing phrase is "local defaults" which I interpreted to refer to cluster-local defaults and not code-local placeholders